[quant] Create nn.quantized.dynamic.EmbeddingBag

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43296 [quant] Add benchmarks for quantized embeddingbag module
* #43176 [quant] Enable from_float for quantized Embedding_Bag
* #43090 [quant] Make offsets an optional argument
* **#43088 [quant] Create nn.quantized.dynamic.EmbeddingBag**

Summary:
Create quantized module that the user can use to perform embedding bag quantization
The module uses the EmbeddingPackedParams to store the weights which can be serialized /deserialized
using TorchBind custom classes (C++ get/setstate code)
Following PR will add support for `from_float` to convert from float to quantized module

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule.test_embedding_bag_api

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23167519](https://our.internmc.facebook.com/intern/diff/D23167519)

supriyar committed:
[quant] Create nn.quantized.dynamic.EmbeddingBag

Summary:
Create quantized module that the user can use to perform embedding bag quantization
The module uses the EmbeddingPackedParams to store the weights which can be serialized /deserialized
using TorchBind custom classes (C++ get/setstate code)
Following PR will add support for `from_float` to convert from float to quantized module

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule.test_embedding_bag_api

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Create nn.quantized.dynamic.EmbeddingBag"


Summary:
Create quantized module that the user can use to perform embedding bag quantization
The module uses the EmbeddingPackedParams to store the weights which can be serialized /deserialized
using TorchBind custom classes (C++ get/setstate code)
Following PR will add support for `from_float` to convert from float to quantized module

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule.test_embedding_bag_api

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Create nn.quantized.dynamic.EmbeddingBag"


Summary:
Create quantized module that the user can use to perform embedding bag quantization
The module uses the EmbeddingPackedParams to store the weights which can be serialized /deserialized
using TorchBind custom classes (C++ get/setstate code)
Following PR will add support for `from_float` to convert from float to quantized module

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule.test_embedding_bag_api

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23167519](https://our.internmc.facebook.com/intern/diff/D23167519)

[ghstack-poisoned]

vkuzo commented on test/quantization/test_quantized_module.py:
```diff
@@ -916,5 +916,63 @@ def test_cell_api(self, dtype):
+        module_out = qemb(indices, offsets)
+
+        # Call the qembedding_bag operator directly
+        ref = torch.ops.quantized.embedding_bag_byte(w_packed, indices, offsets, mode=0,
```
optional: do we need to check other modes as well?  Not sure if that's not needed / done elsewhere / etc

supriyar replied:
We currently don't have support for 4-bit qtensor. Once we add that then we can repeat this check for that op as well.

vkuzo commented on torch/_utils.py:
```diff
@@ -184,11 +184,15 @@ def _rebuild_qtensor(storage, storage_offset, size, stride, quantizer_params, re
         _, scale, zero_point = quantizer_params
         tensor = torch._empty_affine_quantized(size, scale=scale, zero_point=zero_point, dtype=storage.dtype)
-    elif qscheme == torch.per_channel_affine:
+    elif qscheme == torch.per_channel_affine or qscheme == torch.per_channel_affine_float_qparams:
```
nit: `qscheme in (a, b)` ?

vkuzo commented on torch/nn/quantized/dynamic/modules/embeddingbag.py:
```diff
@@ -0,0 +1,147 @@
+        else:
+            raise RuntimeError('Unsupported dtype on dynamic quantized embedding_bag!')
+
+    def forward(self, x):
```
is this just to satisfy the contract of `nn.Module`?

supriyar replied:
Yes

vkuzo commented on torch/nn/quantized/dynamic/modules/embeddingbag.py:
```diff
@@ -0,0 +1,147 @@
+
+        if _weight is None:
+            scales = torch.ones(num_embeddings, dtype=torch.float)
+            zero_points = torch.ones(num_embeddings, dtype=torch.float)
```
optional: should this be zeros?

supriyar replied:
In this case it should be fine because we compute and store the bias as `zero_points * scale * -1` so keeping it non-zero just makes sure we test pre-pack of some actual values.

vkuzo commented on torch/tensor.py:
```diff
@@ -54,7 +54,7 @@ def __deepcopy__(self, memo):
                     if self.qscheme() == torch.per_tensor_affine:
                         quantizer_params = self.qscheme(), self.q_scale(), self.q_zero_point()
-                    elif self.qscheme() == torch.per_channel_affine:
+                    elif self.qscheme() == torch.per_channel_affine or self.qscheme() == torch.per_channel_affine_float_qparams:
```
`in (a, b)`?

vkuzo commented on torch/tensor.py:
```diff
@@ -105,7 +105,7 @@ def __reduce_ex__(self, proto):
                                     self.q_scale(),
                                     self.q_zero_point())
-            elif self.qscheme() == torch.per_channel_affine:
+            elif self.qscheme() == torch.per_channel_affine or self.qscheme() == torch.per_channel_affine_float_qparams:
```
same

vkuzo reviewed (approved):
lg!

supriyar committed:
Update on "[quant] Create nn.quantized.dynamic.EmbeddingBag"


Summary:
Create quantized module that the user can use to perform embedding bag quantization
The module uses the EmbeddingPackedParams to store the weights which can be serialized /deserialized
using TorchBind custom classes (C++ get/setstate code)
Following PR will add support for `from_float` to convert from float to quantized module

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule.test_embedding_bag_api

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23167519](https://our.internmc.facebook.com/intern/diff/D23167519)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Create nn.quantized.dynamic.EmbeddingBag"


Summary:
Create quantized module that the user can use to perform embedding bag quantization
The module uses the EmbeddingPackedParams to store the weights which can be serialized /deserialized
using TorchBind custom classes (C++ get/setstate code)
Following PR will add support for `from_float` to convert from float to quantized module

Test Plan:
python test/test_quantization.py TestDynamicQuantizedModule.test_embedding_bag_api

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23167519](https://our.internmc.facebook.com/intern/diff/D23167519)

[ghstack-poisoned]
